### DISCUSSION TITLE
[RFC] Server-side Website

### DISCUSSION THREAD

**Discussion by mnapoli** [getlift/lift#43] [Constructs]
The goal of this discussion is to get feedback on a "Server-side Website" component.

By *Server-side website*, I mean a website where the HTML is rendered on the server. This is usually done with backend frameworks like Laravel/Symfony (PHP), Ruby on Rails, Django/Flask (Python), Express (Node), etc.

To build a SPA or static website, use the [Static Website construct instead](https://github.com/getlift/lift/blob/master/docs/static-website.md).

## Name

Possible names for the constructs could be:

- server-side website
- server website
- dynamic website
- backend website

Any opinion? IMO "server-side" (or "server") goes nicely with "static website".

## Use case

Deploying websites served via code on Lambda. E.g. websites built with Laravel/Symfony (PHP with Bref), Django/Flask (Python with the [wsgi plugin](https://github.com/logandk/serverless-wsgi)), Express/... (Node), etc.

## Quick start

```yaml
# serverless.yml
# [...]

functions:
    home:
        handler: home.handler
        events:
            -   httpApi: 'GET /'
    # ...

constructs:
    website:
        type: server-side-website
        assets:
            '/css/*': dist/css
            '/js/*': dist/js
```

The example above will set up CloudFront to serve both:

- `https://<domain>/*` -> the website through API Gateway + Lambda
- `https://<domain>/css/*` and `https://<domain>/js/*` ->assets through S3

On `serverless deploy`, assets are uploaded to S3 and the CloudFront cache is invalidated.

_Note: this component takes about 5 minutes to deploy the first time._

## How it works

![image](https://user-images.githubusercontent.com/720328/122544500-27840000-d02d-11eb-9bce-6f13932da350.png)

**Why CloudFront?**

API Gateway's HTTP APIs do not support HTTP. For a website, this is problematic because it means someone typing `website.com` in a browser will get a blank page: API Gateway will not even redirect this to HTTPS.

We need CloudFront to redirect HTTP -> HTTPS.

CloudFront also provides caching of assets (just like for [static websites](https://github.com/getlift/lift/blob/master/docs/static-website.md), as well as routing URLs based on prefixes:

- assets will be served by S3
- all other pages will be served by API Gateway

### CloudFront configuration

CloudFront is configured to cache static assets by default, but not cache dynamic content by default.

CloudFront is configured to forward all headers, cookies, and query strings to the backend running on Lambda.

Using CloudFront Functions, the `X-Forwarded-Host` header will also be automatically populated to resolve a shortcoming of CloudFront + API Gateway. Code running on Lambda will be able to access the original `Host` header via this `X-Forwarded-Host` header.

## Commands

`serverless deploy` deploys everything configured in `serverless.yml` and uploads assets.

It is possible to skip the CloudFormation deployment and directly publish assets changes via:

```
serverless <construct-name>:assets:upload

# For example:
serverless website:assets:upload
```

This command only takes seconds: it directly uploads files to S3 and clears the CloudFront cache.

### API Gateway

```yaml
functions:
    v1:
        handler: foo.handler
        events:
            -   http: 'GET /' # REST API
    v2:
        handler: bar.handler
        events:
            -   httpApi: 'GET /' # HTTP API

constructs:
    website:
        type: server-side-website
        apiGateway: "http" # or "rest"
```

- if `severless.yml` only deploys a API Gateway v1 REST API, then the construct should use it
- if `severless.yml` only deploys a API Gateway v2 HTTP API, then the construct should use it
- if `severless.yml` deploys both a REST API and an HTTP API, then the user should configure which one to use:

If there is no Lambda function configured to respond to HTTP requests then the construct should show an error to the user (an API Gateway is required).

### Assets

```yaml
constructs:
    website:
        # ...
        assets:
            '/assets/*': dist/
```

The `assets` section lets users define routing for static assets (like JavaScript, CSS, images, etc.).

- The key defines **the URL pattern**.
- The value defines **the local path to upload**.

Assets can be either whole directories, or single files:

```yaml
constructs:
    website:
        # ...
        assets:
            # Directories: routes must end with `/*`
            '/css/*': dist/css
            '/images/*': assets/animations
            # Files:
            '/favicon.ico': public/favicon.ico
```

With the example above:

- `https://<domain>/*` -> Lambda
- `https://<domain>/css/*` -> serves the files uploaded from the local `dist/css` directory
- `https://<domain>/images/*` -> serves the files uploaded from the local `assets/animations` directory
- `https://<domain>/favicon.ico` -> serves the file uploaded from `public/favicon.ico`

### Custom domain

```yaml
constructs:
    website:
        # ...
        domain: mywebsite.com
        # ARN of an ACM certificate for the domain, registered in us-east-1
        certificate: arn:aws:acm:us-east-1:123456615250:certificate/0a28e63d-d3a9-4578-9f8b-14347bfe8123
```

This will behave similarly to the [static website domain configuration](https://github.com/getlift/lift/blob/master/docs/static-website.md#custom-domain).

### Error pages

```yaml
constructs:
    website:
        # ...
        errorPage: error500.html
```

In case a web page throws an error, API Gateway's default `Internal Error` blank page shows up. This can be overridden by providing an HTML error page.

Applications are of course free to catch errors and display custom error pages. However, sometimes even error pages and frameworks fail completely: this is where the API Gateway error page shows up.

## Draft pull request

A pull request with a draft implementation is available here: https://github.com/getlift/lift/pull/44

**Comment by t-richard** [comment]
First, this is HUGE ! This is a major hassle for classic fullstack websites when running in Lambda.

> Possible names for the constructs could be:
server-side website
server website
dynamic website
backend website

I would go with the 5th option : Fullstack Website :sweat_smile: 
Backend website could do the trick too I think.

What I dislike about `Server-side website` is that it could be misleading users to think this is a component to do Server-Side Rendering with their favorite frontend framework, especially in the JS community I suppose. (By the way, a construct to support this kind of SSR could be great too).

> CloudFront is configured to cache static assets by default, but not cache dynamic content by default.

Is it Cloudfront edge cache where an invalidation will clear everything up, or is it a cache-control header added to the assets ?

If it's the later, would it be possible to configure which paths are cached and which are not ? 

If using something like webpack, you can add a hash to assets but if your are building something simple, then you wmight not version your assets and some third parties might not allow that at all (for example, Symfony assets provided by third-party bundles are not versionned and I'm not sure you can do it).

Something like 

```yaml
constructs:
    website:
        # ...
        assets:
            '/bundles/*': 
                   path: public/bundles/
                   cache: false # this disables cache-control headers or sets it to 0
            '/build/*': public/build/ # this uses the default cache behaviour
```

Let me know what you think.

**Reply by mnapoli** [reply]
Cool!

To address the points in order:

I'm not convinced by "fullstack website", as it could also be interpreted as a SPA + API website. It's also an uncommon term, searching for this expression on google isn't reassuring: https://www.google.com/search?q="fullstack+website" (searching for "server-side website" gives something a bit more meaningful)

But I do see your point with "SSR" confusion, that's a good one.

> Is it Cloudfront edge cache where an invalidation will clear everything up, or is it a cache-control header added to the assets ?

The idea is to clear the whole cache.

That way it wouldn't matter if assets are prefixed or not, they would be cleared either way. Is that good for you?

**Reply by t-richard** [reply]
> I'm not convinced by "fullstack website"

Not sure about the name either, but I'm sure about the confusion it would cause if I said "Look there is a Server-Side Website construct in Lift" to a frontend colleague :sweat_smile:  
Maybe Backend Rendered Website would be better. Not sure about that one too ^^

If no cache-control header is set, browsers will fallback to a cache duration of 10% the time elapsed since `Last-Modified` header (which would the upload to S3 date in our case). See this [Stack Exchange answer](https://webmasters.stackexchange.com/a/111299) for more details.

So invalidating Cloudfront cache might not be enough if Cache-Control is not set to max-age=0 when uploading to S3. 

From what I can see in the implementation PR, you are looping over each key in `assets` and for each you sync/put to the s3 bucket separately. I think adding cache-control there would be possible based on a config similar to what I suggested above.  
This would also make this construct even more valuable IMHO.

And this proves it again :joy: 
> There are only two hard things in Computer Science: cache invalidation and naming things.

**Reply by mnapoli** [reply]
> If no cache-control header is set, browsers will fallback to a cache duration of 10% the time elapsed since Last-Modified header (which would the upload to S3 date in our case). See this Stack Exchange answer for more details.

🤯  omg this is crazy. Why! 😄 

OK that's a very good point, I wonder if this couldn't cause issues with static websites too 🤔 

Yes! Setting cache headers on S3 objects is definitely something I want to do eventually, for static websites too. But at first I'd love to get something robust working, even if it's not the extremely optimized version. But we need to think about that browser heuristic cache then…

**Reply by t-richard** [reply]
> :exploding_head:  omg this is crazy. Why! :smile: 

Was also shoked when I discovered it. When a site is not updated on a regular basis, browsers could keep some files in cache for days or weeks.

And on the naming again, I think my "Server-Side Rendering could be confusing" point doesn't stand actually. Because I think this construct could also be used on these.

The basis when doing SSR on Lambda is that assets are redirected to S3 by Cloudfront and everything else is passed to a Lambda in charge of the Server-Side Rendering (with Nuxt, Next, etc). So the current approach could work for those too. I guess.

**Reply by mnapoli** [reply]
> If no cache-control header is set, browsers will fallback to a cache duration of 10% the time elapsed since Last-Modified header (which would the upload to S3 date in our case).

@t-richard I'm wrapping up the server-side website, I'm wondering about that: I'm thinking about adding systematically a `cache-control` header that says "don't cache" (so that browsers will never cache the files). But then CloudFront will also not cache them, right?

Note that for now I want to stick to a simple strategy:

- cache in cloudfront
- no browser cache

This is so that it works out of the box. Then we can iterate later on browser caching with more options.

**Reply by t-richard** [reply]
You can control the minimum TTL of Cloudfront cache that will take over the cache-control if `max-age=0`.

In our case we could have `Cache-Control: max-age=0` set on S3 objects and set a Minimum Cache TTL of a given value (24h, 7 days, or more). This way browsers would not cache the file but Cloudfront would use the minimum TTL as a default value for edge caching.

See https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/Expiration.html#ExpirationDownloadDist

![image (1)](https://user-images.githubusercontent.com/22999032/124915148-823bc680-dff1-11eb-9c44-0180204eda3e.png)

It would just mean you'll have to create a custom cache policy to replace [`CachePolicy.CACHING_OPTIMIZED`](https://github.com/getlift/lift/pull/44/files#diff-71c5931026f0ff85c4f1c49a8464c717feaa8fc71f46a80293ab0e622bae0922R373) because it sets the Min TTL to 1s

**Reply by mnapoli** [reply]
So… I'm investigating that again, and I'm questioning the whole thing: I don't see a `Last-Modified` header with CloudFront:

<img width="544" alt="image" src="https://user-images.githubusercontent.com/720328/131229582-5227f8cc-be02-4b42-bab4-b005f24581a4.png">

So maybe we don't have that problem at all here? Was there a specific reason you thought about that header here? I'm afraid of missing something, especially since it's been a while 😅

**Reply by mnapoli** [reply]
Just tested again a real scenario: deployed a CSS file yesterday. I opened the website in my browser and navigated: the CSS was loaded at first and then CloudFront returned 304 (Not Modified).

This morning, I opened the website and navigated it (still 304). Then I uploaded CSS changes (Lift did the CloudFront invalidation). I did not "refresh" the page, simply clicked a link to go to a new page: the new CSS was loaded.

From what I can tell we don't have an issue here.

**Comment by t-richard** [comment]
tested it yesterday with [symfony/demo](https://github.com/symfony/demo) and it worked for the most part.

There are still some hiccups which are blocking it even on a simple app like the demo : 
* `X-Forwarded-Host` is not set so redirects are generated using API Gateway's endpoint (because it's in the `Host` header). It would be nice to set it when the user provides a domain.
* Some headers are whitelisted but you could easily run into cases where more headers are needed (in the case of the demo, it uses `X-Requested-With` to determine if the request is a XHR. Other headers like `X-Api-Key` or CORS related headers. Some custom headers might be needed too so I think it's important to allow configuration of the headers.

Otherwise, it seems to work well. I'll to test this with a Server-Side rendered Vue app I have and see how it turns out.

**Reply by t-richard** [reply]
Also Cloudfront functions to handle www redirects could be a nice addition. I know it was also discussed in the static website RFC.

**Reply by mnapoli** [reply]
Awesome, thank you for testing it!

I was already planning on forwarding the real host header to `X-Forwarded-Host` with CloudFront Functions, because yeah I definitely want to solve that problem once and for all.

It's really annoying that we can't configure CloudFront to forward "all" headers, because we must not forward "Host" else it breaks API Gateway. Making headers configurable is a very good point, I'll do that!

> www redirects

Yes, it would make sense for both. 
I was thinking about this and maybe "generic" redirects could make sense too, a bit like https://docs.netlify.com/routing/redirects/redirect-options/ I mean, as long as we upload CloudFront Functions for redirection logic, we might as well support more cases? Not 100% sure about that yet, that's food for thoughts.

**Reply by t-richard** [reply]
> I was already planning on forwarding the real host header to X-Forwarded-Host with CloudFront Functions

Why not just use `OriginCustomHeaders` to do that ? Trusting the Host header can be a security issue in some cases. But maybe that's not a problem with cloudfront because passing a Host that is not whitelisted on the the distribution fails :thinking: 

www redirect is the most basic one that needs to be supported IMHO. 

With www redirect, X-Forwarded-Host and custom headers support, that would cover almost all use cases. Most websites use the "same" basic Apache/nginx config anyway.

But redirect rules like that could be awesome in the future :+1: I've already used Netlify with those and they truly are powerful. 
I even used netlify just for that because I needed some redirection rules and it was an easy, cheap and efficient solution to do that :joy: 

But maybe that inside Cloudfront functions could be problematic and very limited. Cloudfront functions must execute under 1ms IIRC so this could cause some issues with large sets of rule.

**Reply by mnapoli** [reply]
Yeah you're making a good point: a server-side website can implement any redirect it wants… Including the `www` one actually. So I'm not even sure we need to support redirects at all?

> Why not just use OriginCustomHeaders to do that ? Trusting the Host header can be a security issue in some cases. But maybe that's not a problem with cloudfront because passing a Host that is not whitelisted on the the distribution fails 🤔

Yeah, I don't see a security issue (but let's be careful here).

What do you have in mind with `OriginCustomHeaders`? What do you mean by that?

**Reply by t-richard** [reply]
> What do you have in mind with OriginCustomHeaders? What do you mean by that?

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-cloudfront-distribution-origin.html#cfn-cloudfront-distribution-origin-origincustomheaders

You can add fixed headers in the cloudformation and add the user provided domain there (I think this is what is outlined in the assets docs of bref for the x-forwarded-header)

> Yeah you're making a good point: a server-side website can implement any redirect it wants… Including the www one actually

I think it makes sense for advanced redirects but www redirects would be great to have by default. It would feel wrong to implement this in the server code IMHO.

**Reply by mnapoli** [reply]
🤦  oh ok I see what you meant now. Yeah hardcoding the header works if you only host your app on a single domain. If you support multiple domains, then you are out of luck. I was thinking of solving this once and for all with CloudFront Functions.

The security aspect is a good point but AFAICT CloudFront will not serve your app on an invalid header (I should test that to be sure), so that shouldn't be an issue here.

**Reply by t-richard** [reply]
> The security aspect is a good point but AFAICT CloudFront will not serve your app on an invalid header (I should test that to be sure), so that shouldn't be an issue here.

I tested it and it is apparently not possible to "poison" the `Host` header with Cloudfront. This was also stated in some articles like [this one](https://www.mindpointgroup.com/blog/cyber-security/cloudfront-hijacking/) which says

> If the HOST header does not match a domain in the “Alternate Domain Names (CNAMEs)” field of the intended distribution, the request will fail.

And the rest of the article is not of interest to us I think even if it's a quite interesting read :nerd_face: 

> If you support multiple domains, then you are out of luck

True, so Cloudfront functions are the way out of this. The pricing and overhead are so low that it seems like the best solution.

**Reply by mnapoli** [reply]
Quick update following some attempts:

- when a single custom domain is set, we can use `OriginCustomHeaders`
- when multiple custom domains are set, we cannot use `OriginCustomHeaders`, we must use CFF
- when no custom domains are set, we cannot put the CloudFront domain in `OriginCustomHeaders` because we cannot do a CloudFormation reference to itself…

All that leads me to think that deploying CFF all the time is a simpler solution.

For reference, here is a possible solution for actually having the correct `Host` header: https://stackoverflow.com/a/53343079/245552 But since this isn't documented, and this is quite a hassle to set up, I don't want to go that route. I think having `X-Forwarded-Host` is acceptable.

**Reply by t-richard** [reply]
I think it makes sense to do it consistently with CFF. It will be easier to implement, explain and maintain.

**Comment by mykiwi** [comment]
Quick question, how can I test it on a project? It is not merged yet, so I'm not sure what should I do to install this package version.

**Reply by t-richard** [reply]
You can test it using `npm i github:getlift/lift#server-side-website` 🙂

**Reply by mykiwi** [reply]
I tried this command, but I got an error :thinking: 
```
❯ cat package.json 
{
  "dependencies": {
    "serverless": "^2.52.1"
  }
}
                                                                                              
❯ npm i github:getlift/lift#server-side-website
npm ERR! prepareGitDep 2> npm WARN install Usage of the `--dev` option is deprecated. Use `--also=dev` instead.
npm ERR! prepareGitDep npm ERR! code ENOTSUP
npm ERR! prepareGitDep npm ERR! notsup Unsupported engine for serverless-lift@: wanted: {"node":">=10 <16"} (current: {"node":"16.4.2","npm":"6.14.8"})
npm ERR! prepareGitDep npm ERR! notsup Not compatible with your version of node/npm: serverless-lift@
npm ERR! prepareGitDep npm ERR! notsup Not compatible with your version of node/npm: serverless-lift@
npm ERR! prepareGitDep npm ERR! notsup Required: {"node":">=10 <16"}
npm ERR! prepareGitDep npm ERR! notsup Actual:   {"npm":"6.14.8","node":"16.4.2"}
npm ERR! prepareGitDep 
npm ERR! prepareGitDep npm ERR! A complete log of this run can be found in:
npm ERR! prepareGitDep npm ERR!     /.../npm/_logs/2021-07-24T09_58_00_973Z-debug.log
npm ERR! prepareGitDep 
npm ERR! premature close

npm ERR! A complete log of this run can be found in:
npm ERR!     /.../.npm/_logs/2021-07-24T09_58_01_004Z-debug.log
```

**Reply by t-richard** [reply]
I think your node version is too high
See https://github.com/getlift/lift/blob/5ad131bc5086588d4388d4f9feb0f5fe933bbba0/package.json#L66

Could you try with node ^15.14 ?

**Reply by mykiwi** [reply]
I will try, but what I don't understand is that it is working correctly with the main version and this requirement is the same :thinking:

**Reply by t-richard** [reply]
When you install from a branch, npm will run the "prepare" script so you need development requirements to run this so that's why it is different than installing a published version which is already built.

Basically npm will clone the repo in a tmp dir, install dependencies (including dev), run the prepare script which will build the project then copy that in node_modules of the project you are trying to install it into. It's because of the build phase that you need different versions.

**Reply by mykiwi** [reply]
This works! Thank you :smiley:

**Reply by mnapoli** [reply]
@mykiwi sorry, holidays here :)

Happy to see some interest in this, let me know if you have any feedback on the construct.

**Comment by mykiwi** [comment]
Thank you for this! This works great.

But how can we correctly handle different stages?
For example when I try to deploy my site with `serverless deploy --stage=foo` this does not works. Here is the output:
```
 Serverless Error ----------------------------------------
 
  An error occurred: websiteCDN012346 - Resource handler returned message: "Invalid request provided: One or more of the CNAMEs you provided are already associated with a different resource. (Service: CloudFront, Status Code: 409, Request ID: abcdef, Extended Request ID: null)" (RequestToken: ghijklm, HandlerErrorCode: InvalidRequest).
 
  Get Support --------------------------------------------
     Docs:          docs.serverless.com
     Bugs:          github.com/serverless/serverless/issues
     Issues:        forum.serverless.com
 
  Your Environment Information ---------------------------
     Operating System:          linux
     Node Version:              15.14.0
     Framework Version:         2.52.1
     Plugin Version:            5.4.3
     SDK Version:               4.2.5
     Components Version:        3.14.0
```

It could be nice to allow a main stage = main domain (`my.domain.com`) & custom stage = a (customizable?) pattern like `{stage}.my.domain.com` or another cloudfront url.

**Reply by t-richard** [reply]
You are able to combine Lift constructs with Serverless framework variables. Something like this should do the trick

```yaml
...

custom:
  domain:
    prod: my.domain.com
    other: ${sls:stage}.my.domain.com

constructs:
    website:
        type: server-side-website
        domain: ${self:custom.domain.${sls:stage}, self:custom.domain.other}
        assets:
            '/css/*': dist/css
            '/js/*': dist/js
```

This way, `serverless deploy --stage=prod` would use `my.domain.com` and all others would use `{stage}.my.domain.com`

You could also use an env variable if you deploy via CI/CD

IMHO, this is out of the scope of Lift because Serverless Framework allows for a lot of flexibility already.

**Reply by mnapoli** [reply]
👍  one solution is to have something like:

```yaml
constructs:
    website:
        # ...
        domain: ${sls:stage}.mywebsite.com
```

but it wouldn't really work for prod (`prod.mywebsite.com` instead of `mywebsite.com`).

Or you could end up with something complex like:

```yaml
custom:
    domains:
        prod: mywebsite.com
        staging: staging.mywebsite.com

constructs:
    website:
        # ...
        domain: ${self:custom.domains.${sls:stage}}
```

I'm not sure if there's any native feature we could include in this construct specifically (what would it look like?). After all, the same problem probably applies to all constructs.

Maybe another approach would be something more generic, like:

```yaml
constructs:
    website:
        if: stage == 'prod'
        # ...
        domain: mywebsite.com

    website-dev:
        if:  stage !== 'prod'
        # ...
        domain: ${sls:stage}.mywebsite.com
```

This kind of logic could apply to all constructs.

**Reply by mykiwi** [reply]
Thanks for the answers! I will try it :smiley:

**Comment by mnapoli** [comment]
Hey all, a quick update. I've been reading about HSTS (which is nothing new) and TIL: there are now lists maintained by Chrome and Firefox of "HTTPS-only" websites (https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Strict-Transport-Security#preloading_strict_transport_security).

That allows a website to be 100% HTTPS-only, and still work in all browsers.

That makes me think that we may want to consider hosting websites without CloudFront at all? CloudFront is still useful to host everything behind a single domain and benefit from CDN caching. But it takes a while to setup/deploy.

Maybe we will want a simple-website construct? I'm just throwing the idea out there. CloudFront may not be absolutely required anymore.

**Reply by piolet** [reply]
a website with a 100% server rendering?
But then, where would the assets be if there is no more CDN?

**Reply by mnapoli** [reply]
Served directly from S3

**Reply by piolet** [reply]
Hum, ok... With static web host option activated ?

But S3 don't manage https, that's not a problem?

**Reply by mnapoli** [reply]
No that would be without static website hosting (so HTTPS is supported). Static websites would not be supported (since this is a server-side website construct). 100% server rendering.

**Reply by piolet** [reply]
ok ok.

Not yet my use case, but ok. It make sense

**Comment by jjmontgo** [comment]
How would I handle a case where I want to provide public access to user uploaded files?  Is there any way to map a path, /uploads, to a storage S3 bucket?

**Reply by mnapoli** [reply]
The bucket created in the `server-side-website` construct is for storing assets. Not for user-uploaded files.

I encourage you to create a separate bucket for the uploaded files.

**Reply by jjmontgo** [reply]
That's the strategy I've taken -- with a separate bucket and a separate cloudfront distribution for accessing a public folder in the bucket.  It would have been nice if I could have added an additional backend to the distribution created by the plugin, but you can't have it all.  Thank you!

**Comment by hmazter** [comment]
Would it be possible to create a new version of this construct or update this to use the Lambda Function Url instead of API Gateway?

Lambda Function Url seems like the perfect companion for this construct together with something like Laravel

**Reply by mnapoli** [reply]
@hmazter we could indeed, but I wonder if adding another option like this one https://github.com/getlift/lift/blob/master/docs/server-side-website.md#api-gateway would make sense? It would be much simpler. For example:

```yaml
constructs:
    website:
        type: server-side-website
        lambdaUrl: function-name
```

WDYT?

**Reply by hmazter** [reply]
It works for me 👍 
Just did not want to suggest increased underlying complexity 😄

**Reply by mnapoli** [reply]
I think for this case specifically it's OK

**Comment by AtteR** [comment]
Hello,

With the new extensions, is it possible to add new origins for website construct? I have tested this:

`website:

    extensions:

      distribution:

        Properties:

          DistributionConfig:

            Origins:

              - DomainName: ${construct:contents.bucketName}.s3.{$aws:region}.amazonaws.com
              
                Id: ContentsOrigin

                S3OriginConfig:

                  OriginAccessIdentity: !Join ['', ['origin-access-identity/cloudfront/', !Ref CloudFrontContentsAi]]
`

Getting this error:

`Invalid request provided: One or more of your origins or origin groups do not exist`

Is my .yml simply wrong or is there something I am missing? Looking to create Cloudfront distribution for user uploaded files. Do I need to create completely new distribution for this?

Thank you.

**Reply by mnapoli** [reply]
I don't think you can do that as it might override the `Origins` entirely?

**Reply by kevincerro** [reply]
I'm trying to do this also.

We want to add one origin and one behaviour to the cloudfront created by lift server-side-construct

Nowadays is possible to do this? @mnapoli

**Reply by mnapoli** [reply]
@kevincerro I don't think it's possible to do this easily, not sure where to point you except to create your own cloudformation template

**Reply by kevincerro** [reply]
@mnapoli Maybe this can be used 
Our use case is to host under a subpath /blog/ a static-website construct / s3 bucket 
https://github.com/getlift/lift/pull/325
